### PR TITLE
fixes cancelling PlayerTradeEvent

### LINKED ISSUE #5263: PlayerTradeEvent cannot be cancelled.
### What behaviour is expected:
___PlayerTradeEvent can be cancelled properly.___

### What behaviour is observed:
___PlayerTradeEvent cancellation has no effect.___

### Steps/models to reproduce:
- Use this EventHandler: image
- Trade with a villager
- See that you can still trade no matter you cancel PlayerTradeEvent or not.

### Plugin list:
___ViaVersion, ViaBackwards and my plugin.___

### Paper version:
___This server is running Paper version git-Paper-499 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT) You are running the latest version.___

### LINKED ISSUE #5552: PlayerTradeEvent
### Is your feature request related to a problem?

Call the PlayerTradeEvent when trading on a CraftMerchantCustom. This will allow me to increment the usage value and check if the maximum number of trades has not been reached.

### Describe the solution you'd like.

Call the PlayerTradeEvent or create another one with a null abstractvillager

### Describe alternatives you've considered.

Allow a merchant to open his trading interface to multiple players and if the maximum usage value has been reached cancel the trading event and close the interface.

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### PR DESCRIPTION
**Pull request by Machine-Maker**
Fixes #5263

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Actually, this has a problem still

**Comment by Machine-Maker** [comment]
Ok yeah, went ahead an added a second event with it's own handler set for custom merchants.

**Comment by electronicboy** [comment]
https://github.com/PaperMC/Paper/pull/5583

**Comment by Machine-Maker** [comment]
Wouldn't that be an even bigger API break than just changing the existing event to take null AbstractVillagers?

**Comment by electronicboy** [comment]
making something nullable which isn't nullable is a behavioral break which is somewhat iffy, also adds support for rewarding exp during a custom trade, which, no idea if that's wanted, but, is a thing; mainly just wanted to make you aware that something already existed and wasn't sure how it compared to this

**Comment by Machine-Maker** [comment]
I mean... so is totally changing where the existing event is called, which is what #5583 is doing. PlayerTradeEvent is having its method signature changed, and where it is called changed. Its probably the way I should've done it in the first place for sure, but that seems like a bigger behavioral break, whereas this shouldn't break any existing code anywhere in some plugin I dont think.

**Comment by electronicboy** [comment]
Moving methods up the class hierarchy won't break ABI last I recall, I may be wrong there, but, fairly sure on that

**Comment by Machine-Maker** [comment]
Reworked to just fix #5263. Probably should be blocked until #5583 is merged because this pr needs to be adjusted for that change.

**Review by electronicboy — APPROVED** [review]

File: Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch]: obfhelper

File: Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch]: You don't need to keep it all on one line: do proper line breaks so this is readable

File: Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch]: Add the fact this is reward XP.

File: Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch]: Why are these two blocks?

**Review by Proximyst — APPROVED** [review]
Minor changes necessary

File: Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch
Comment (Machine-Maker) [reply Spigot-Server-Patches/0619-added-PlayerTradeEvent.patch]: Because as soon as #5583 is merged, there is gonna be an `else if` right after that to construct an event that extends PlayerTradeEvent. so yeah, right now it looks a little strange, but it'll make sense after that other pr is merged.

**Comment by Machine-Maker** [comment]
Implemented changes

**Review by Proximyst — APPROVED** [review]

**Comment by Machine-Maker** [comment]
Remember... **dont merge this yet**, maybe the blocked label would be good here?

**Comment by Proximyst** [comment]
Could you base this on top of your other PR so both commits show up? That should make what it's blocked on fairly obvious.

**Comment by Machine-Maker** [comment]
Other PR isnt mine, but I guess I can checkout their branch

**Comment by Machine-Maker** [comment]
Ok, this also now moves the firing of PlayerPurchaseEvent into the same area of the patch history as the api patch where its added.

**Comment by Machine-Maker** [comment]
This can be unblocked btw, the blocking PR was merged earlier today

**Review by jpenilla — DISMISSED** [review]
It seems like cancellation does not work properly when shift-clicking the result item.

When I click normally, it seems to work fine, the payment is not taken and the reward is not given, although there is some client desync with the villager experience which is fixed when reopening the GUI (which is fine).

The issue is that when shift clicking the result, the payment doesn't get taken and the villager doesn't increase experience (as expected), however my entire inventory filled up with the result item. Testing was done in survival mode.

**Comment by Machine-Maker** [comment]
Huh. I guess I didn’t think there would be a difference. I’ll have to go back and find out what’s different there. And yeah, I don’t know any way around the xp desync, there may be one, but idk.

**Comment by Machine-Maker** [comment]
If someone sees a simple fix for this, please let me know. What I see, is in the overridden shiftClick method in ContainerMerchant, the ItemStack is moved well before the event is even called. I can refactor that to only check if the item could be moved, but that’s a much bigger job.

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

**Review by ArnaudLier — APPROVED** [review]

**Pull request merged**
